rdata.c

```diff
@@ -67,414 +69,455 @@ lookup_table_type dns_algorithms[] = {
 		if (isprint((unsigned char)ch)) {
 			if (ch == '"' || ch == '\\' || ch == '(' || ch == ')'
 			  || ch == '\'' || isspace((unsigned char)ch)) {
 				buffer_printf(output, "\\");
 			}
 			buffer_printf(output, "%c", ch);
 		} else {
-			buffer_printf(output, "\\%03u", (unsigned) data[i]);
-		}
-	}
-	return 1;
-}
-
-static int
-rdata_unquoteds_to_string(buffer_type *output, rdata_atom_type rdata,
-	rr_type* ATTR_UNUSED(rr))
-{
-	uint16_t pos = 0;
-	const uint8_t *data = rdata_atom_data(rdata);
-	uint16_t length = rdata_atom_size(rdata);
-	size_t i;
-
-	while (pos < length && pos + data[pos] < length) {
-		for (i = 1; i <= data[pos]; ++i) {
-			char ch = (char) data[pos + i];
-			if (isprint((unsigned char)ch)) {
-				if (ch == '"' || ch == '\\'
-				||  isspace((unsigned char)ch)) {
-					buffer_printf(output, "\\");
-				}
-				buffer_printf(output, "%c", ch);
-			} else {
-				buffer_printf(output, "\\%03u", (unsigned) data[pos+i]);
-			}
+			buffer_printf(output, "\\%03u",
+				(unsigned) rdata[*offset + i]);
 		}
-		pos += data[pos]+1;
-		buffer_printf(output, pos < length?" ":"");
 	}
+	*offset += 1;
+	*offset += len;
 	return 1;
 }
 
 static int
-rdata_tag_to_string(buffer_type *output, rdata_atom_type rdata,
-	rr_type* ATTR_UNUSED(rr))
+print_unquoteds(buffer_type *output, uint16_t rdlength,
+	const uint8_t* rdata, uint16_t* offset)
 {
-	const uint8_t *data = rdata_atom_data(rdata);
-	uint8_t length = data[0];
-	size_t i;
-	for (i = 1; i <= length; ++i) {
-		char ch = (char) data[i];
-		if (isdigit((unsigned char)ch) || islower((unsigned char)ch))
-			buffer_printf(output, "%c", ch);
-		else	return 0;
+	while (*offset < rdlength) {
+		if(!print_unquoted(output, rdlength, rdata, offset))
+			return 0;
+		if(*offset < rdlength)
+			buffer_printf(output, " ");
 	}
 	return 1;
 }
 
+/*
+ * Print IP4 address.
+ * @param output: the string is output here.
+ * @param rdlength: length of rdata.
+ * @param rdata: the rdata. The rdata+*offset is where the field is.
+ * @param offset: the current position on input. The position is updated to
+ *	be incremented with the length of rdata that was used.
+ * @return false on failure.
+ */
 static int
-rdata_byte_to_string(buffer_type *output, rdata_atom_type rdata,
-	rr_type* ATTR_UNUSED(rr))
-{
-	uint8_t data = *rdata_atom_data(rdata);
-	buffer_printf(output, "%lu", (unsigned long) data);
-	return 1;
-}
-
-static int
-rdata_short_to_string(buffer_type *output, rdata_atom_type rdata,
-	rr_type* ATTR_UNUSED(rr))
-{
-	uint16_t data = read_uint16(rdata_atom_data(rdata));
-	buffer_printf(output, "%lu", (unsigned long) data);
-	return 1;
-}
-
-static int
-rdata_long_to_string(buffer_type *output, rdata_atom_type rdata,
-	rr_type* ATTR_UNUSED(rr))
+print_ip4(struct buffer *output, size_t rdlength, const uint8_t *rdata,
+	uint16_t *offset)
 {
-	uint32_t data = read_uint32(rdata_atom_data(rdata));
-	buffer_printf(output, "%lu", (unsigned long) data);
+	char str[INET_ADDRSTRLEN + 1];
```

Comment:
Do you want to add the assertion here too, like in the other print functions below?
```suggestion
	char str[INET_ADDRSTRLEN + 1];
	assert(rdlength >= *offset);
```

Reply:
Certainly, the assertions can prove useful. And similarity with the other functions is good too.